pkgs/by-name/qm/qmplay2/sources.nix

```diff
@@ -0,0 +1,21 @@
+{
+  fetchFromGitHub,
+}:
+
+{
+  qmplay2 = let
+    self = {
+      pname = "qmplay2";
+      version = "24.06.16";
+
+      src = fetchFromGitHub {
+        owner = "zaps166";
+        repo = "QMPlay2";
+        rev = self.version;
+        fetchSubmodules = true;
+        hash = "sha256-ECDEyF16GwIKn1raG7J56HFMYk4G6ezwAG2BEy8hUZs=";
+      };
+    };
+  in
+    self;
+}
```

Comment:
The changes seem reasonable, but I'm having a hard time wrapping my head around the reasoning behind this block. Do you mind if I ask?

How does it help that the code is put into a seperate file if `fetchSubmodules` become discouraged or deprecated in the future?

And what is up with the weirdly recursive let block? Does it somehow make it overridable as a fixpoint expression from `pkgs.qmplay2`?

Reply:
> How does it help that the code is put into a seperate file if fetchSubmodules become discouraged or deprecated in the future?

As far as I rememeber, `fetchSubmodules` has some problems with reproducibility.
My intention is to push those submodules directly in a near future.

But this would generate a big block of code that is just a truckload of fetchers.
It clutters the already big package file.
So I detached the source acquisition to a separated file.

It indeed can help maintenance, since `package.nix` will deal with the build proper and `sources.nix` will deal with sources acquisition.

I have made a similar split before:

https://github.com/NixOS/nixpkgs/pull/310831
https://github.com/NixOS/nixpkgs/pull/319877
https://github.com/NixOS/nixpkgs/pull/322093

> And what is up with the weirdly recursive let block? 

I do not like `rec`. It caused too much problem before.
I prefer let-in, since it shows clearly the origin of each attribute.

Reply:
Sorry, I think I worded my questions a bit poorly. What I meant to ask was:

- I can see why you would want to split the source fetchers off the package once they become large, but is there any merit to doing it preemptively?

- Avoiding `rec` is fair, but why use recursion within the `let` block? Wouldn't it be easier to both read and write (and possibly evaluate) something like this instead?

```nix
qmplay2 = 
  let
    version = "24.06.16";
  in
  {
    pname = "qmplay2";
    inherit version;

    src = fetchFromGitHub {
      owner = "zaps166";
      repo = "QMPlay2";
      rev = version;
      fetchSubmodules = true;
      hash = "sha256-ECDEyF16GwIKn1raG7J56HFMYk4G6ezwAG2BEy8hUZs=";
    };
  };
```

Reply:
> I can see why you would want to split the source fetchers off the package once they become large, but is there any merit to doing it preemptively?

1. Yes, I think it does. Not a huge merit, but a non-negligible one. 
2. Indeed I got rid of it right now.

> Wouldn't it be easier to both read and write (and possibly evaluate) something like this instead?

This is a matter of personal taste. 
`let-self-in-self` is a bit more typical in Nixpkgs, and I believe it has a clearer intention than merely factoring equal terms. In other words, it is a `rec` without the problems of `rec`.